Preliminary implementation of digesting mobile device regexes.

peteygao opened the pull request:
I updated the documentations to reflect this change (I also arbitrarily incremented the version to `0.4.0`, but you're free to control the versioning).

This uses the updated/fixed `mobiles.yml` (the fixes which I have a PR open for in the original repo [https://github.com/piwik/device-detector/pull/5313](here))

A test usage:
```ruby
ua = "Mozilla/5.0 (iPhone; CPU iPhone OS 8_1_3 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Mobile/12B466 WindVane tae_sdk_ios_1.3.0 Weibo (iPhone5,2__weibo__5.1.0__iphone__os8.1.3)"

device = DeviceDetector.new ua
# The two lines below are New :)!
device.device_name #=> 'iPhone 5'
device.device_type #=> 'smartphone'
# The lines below are the same as the old/standard behaviour
device.os_name #=> 'iOS'
device.os_full_version #=> '8_1_3'
```

yagooar commented:
Could you please provide some specs for the PR? The regex loading part seems critical to me. The new public methods `#device_name` and `#device_type` need to be tested, too.

Thanks!

peteygao commented:
Created tests for the new functionalities I added. I also created a DeviceExtractor class to extract dynamic model names (much like VersionExtractor for versions))

The only thing I did not test is a direct call to the private class method `self#recursive_regex_parser` as that's an anti-pattern (http://blog.rubybestpractices.com/posts/gregory/034-issue-5-testing-antipatterns.html). It would tie the test too closely to the implementation.

Instead, I have implemented a test which tests the `#name` and `#device_type` calls on different types of nested YAML hierarchies, which will allow some future developer to swap out the `self#recursive_regex_parser` implementation should the need arise.

yagooar commented:
Yes, by testing I meant a more high-level approach on the public interface. Going to review the code in the next hour, be right back.

yagooar commented on lib/device_detector/device.rb:
```diff
@@ -0,0 +1,54 @@
+      ModelExtractor.new(user_agent, regex_meta).call
+    end
+
+    def device_type
```
Why not just call the method `type`? Being in the context of `Device` it shall be clear which `type` it is.

yagooar commented on lib/device_detector/device.rb:
```diff
@@ -0,0 +1,54 @@
+    def self.regexes_for(filepaths)
+      @regexes ||=
+        begin
+          regexes = YAML.load(filepaths.map { |filepath| File.read(filepath) }.join)
```
Could you please make two lines out of this one? The `filepaths` joining would be more clear assigned to a separate variable.

yagooar commented on lib/device_detector/device.rb:
```diff
@@ -0,0 +1,54 @@
+        begin
+          regexes = YAML.load(filepaths.map { |filepath| File.read(filepath) }.join)
+
+          recursive_regex_parser(regexes)
```
I have an idea on how to avoid overriding the whole `.regexes_for(:filepaths)` method.

In `lib/device_detector/parser.rb` we could rewrite the method like this:

```ruby
    def self.regexes_for(filepaths)
      @regexes ||=
        begin
          regexes = YAML.load(filepaths.map { |filepath| File.read(filepath) }.join)
          parse_regexes(regexes)
        end
    end

    def self.parse_regexes(regexes)
        regexes.map do |meta|
          meta['regex'] = Regexp.new(meta['regex'])
          meta
        end
    end
```

This would allow to just override the `.parse_regexes(regexes)` method.

yagooar commented on lib/device_detector/device.rb:
```diff
@@ -0,0 +1,54 @@
+    def self.recursive_regex_parser(nested_regex)
+      nested_regex.map { |base, nest|
+
+        if !nest.nil? && nest.include?('models')
```
Would `nest.key?('models')` be an option here?

yagooar commented on lib/device_detector/device.rb:
```diff
@@ -0,0 +1,54 @@
+        if !nest.nil? && nest.include?('models')
+          recursive_regex_parser nest['models']
+        else
+          case base.class.to_s
```
You could simply rewrite the case like this:

```ruby
case base
  when Hash
    base['regex'] = Regexp.new base['regex']
    base
  else
    nest['regex'] = Regexp.new nest['regex']
    nest
end
```

yagooar replied:
Please also use the format with `case` and `when` statements with same identation.

yagooar commented on lib/device_detector/device.rb:
```diff
@@ -0,0 +1,54 @@
+            when 'Hash'
+              base['regex'] = Regexp.new base['regex']
+              base
+            else
```
Probably there is no need for type checking in here, but what if `base` is something unexpected? Maybe it make the code more solid if there was an explicit check, and fail if input is unacceptable:

```ruby
case base
  when Hash
    base['regex'] = Regexp.new base['regex']
    base
  when String # pseudo
    nest['regex'] = Regexp.new nest['regex']
    nest
  else
    fail('Base cannot be Foo')
end
```

yagooar commented on lib/device_detector/model_extractor.rb:
```diff
@@ -0,0 +1,11 @@
+class DeviceDetector
+  class ModelExtractor < VersionExtractor
```
This is a bit weird, as it somehow implies that a model is a subtype of version. How about building an abstract `MetadataExtractor`, which both `VersionExtractor` and `ModelExtractor` could inherit from? It so, the method `version_string` would have to be renamed to something more generic.

peteygao replied:
This refactor has been made.

yagooar commented on lib/device_detector/version_extractor.rb:
```diff
@@ -18,13 +18,17 @@ def extract_version
 
     def replace_version_string_with(match_data)
-      string = version_string
+      string = version_string.to_s
```
Shouldn't this be better done in the `#version_string` method?

You could use the following pattern:

```ruby
def version_string
  String(regex_meta['version'])
end
```

yagooar commented on lib/device_detector/version_extractor.rb:
```diff
@@ -18,13 +18,17 @@ def extract_version
+          string = string.gsub(/\$#{index}/, match_data[index])
+        else
+          string = string.gsub(/\$#{index}/, '')
+        end
```
No need here for duplicating the logic:

```ruby
matched_data = String(match_data[index])
string = string.gsub(/\$#{index}/, matched_data)
```

yagooar commented on lib/device_detector/version_extractor.rb:
```diff
@@ -18,13 +18,17 @@ def extract_version
       end
 
-      string
+      string.strip
```
Could you explain how / when could it happen, that the string has some trailining whitespace in it?

peteygao replied:
Yes, for regexes like:
```yaml
- regex: '(?:Apple-)?iPhone ?(3GS?|4S?|5[CS]?|6(:? Plus)?)?'
  model: 'iPhone $1'
  device: 'smartphone'
```

If the first match is made but nothing for the capture group, the output will be `'iPhone '` (empty trailing space at the end).

yagooar commented on spec/device_detector/device_spec.rb:
```diff
@@ -0,0 +1,48 @@
+require 'spec_helper'
+
+describe DeviceDetector::Device do
```
Could you please provide at least one example for `nil` cases when calling `#name` and `#device_type`?

peteygao replied:
Yep, test added.

yagooar replied:
:+1:

yagooar commented on spec/device_detector/device_spec.rb:
```diff
@@ -0,0 +1,48 @@
+    context 'when models are nested' do
+      let(:user_agent) { 'Mozilla/5.0 (iPhone; CPU iPhone OS 8_1_3 like Mac OS X) AppleWebKit/600.1.4 (KHTML, like Gecko) Mobile/12B466 [FBDV/iPhone7,2]' }
+
+      it 'should find an Apple iPhone 6' do
```
A bit stylistic thing, I like the expectations to be expressing themselves in a "confident" way, e.g `it 'finds an Apple iPhone6`, without the should.

yagooar commented on spec/device_detector/model_extractor_spec.rb:
```diff
@@ -0,0 +1,61 @@
+        let(:device_name) { 'iPhone' }
+
+        it 'should return the textual portion without trailing whitespace' do
+          expect(extractor.call).to eq(device_name)
```
How about putting explicitly `iPhone` without the `let`? It would make it clearer as what to expect, plus it is only used in this line, actually.

peteygao replied:
I like to have the device name as a `let` for consistency, and if we ever add more tests + have to change the device name, the changes can be made in one place :).

yagooar commented on spec/device_detector/version_extractor_spec.rb:
```diff
@@ -36,7 +37,12 @@
+      end
+
+      it 'should remove trailing white spaces' do
+        regex_meta['version'] = regex_meta['version'] + '   ' # 3 trailing spaces
```
You could assign this line to a variable called `version_with_trailing_spaces`, and than you wouldn't need the comment ;) This makes for better maintenance of the code, as the variable name is unlikely to get out of sync, as the comment does.

yagooar commented:
Thank you very much for the PR. It adds an important piece of utility to the gem. I have commented some stuff here and there, but the overall quality of it is very high!

So please take your time to work in the feedback, see you in the next round ;)

yagooar commented on lib/device_detector/metadata_extractor.rb:
```diff
@@ -0,0 +1,37 @@
+    private
+
+    def metadata_string
+      fail "#{self.name} (a child of MetadataExtractor) must implement the '#{__method__}' method."
```
Could you please make this fail with a `NotImplementedError`?

```ruby
message = "#{self.name} (a child of MetadataExtractor) must implement the '#{__method__}' method."
fail NotImplementedError, message
```

yagooar commented on lib/device_detector/metadata_extractor.rb:
```diff
@@ -0,0 +1,37 @@
+      fail "#{self.name} (a child of MetadataExtractor) must implement the '#{__method__}' method."
+    end
+
+    def extract_version
```
I guess we need to rename this method and `replace_version_string_with`. Ideas for a better name?

yagooar commented:
We are getting there! Probably one last round and we're good to merge ;)

I'm going to release a new version as soon as this is merged.

peteygao commented:
Hey, just pinging you to mention that the refactors have been made and are awaiting review :).

yagooar commented:
Thanks for the ping, just reserved a slot tomorrow for the review.

yagooar commented:
Everything works perfectly. :+1:  Thanks a lot for the PR! Going to release a new version in a few minutes.
